src/Store/PaneReducer.re

```diff
@@ -0,0 +1,48 @@
+/*
+ * PaneReducer.re
+ */
+
+open Oni_Model;
+
+let reduce = (action: Actions.t, state: State.t) => {
+  let {pane, _}: State.t = state;
+  let pt = Pane.getType(state.pane);
+
+  let isSearchOpen = pt == Some(Pane.Search);
+  let isSearchFocused = FocusManager.current(state) == Search;
+  let isDiagnosticsOpen = pt == Some(Pane.Diagnostics);
+
+  switch (action) {
+  | SearchHotkey when isSearchOpen && isSearchFocused =>
+    let ret =
+      {...state, pane: Pane.setClosed(pane)} |> FocusManager.pop(Search);
+    ret;
+  | ActivityBar(ActivityBar.SearchClick) when isSearchOpen && isSearchFocused =>
+    {...state, pane: Pane.setClosed(pane)} |> FocusManager.pop(Search)
+  | SearchHotkey when isSearchOpen && !isSearchFocused =>
+    state |> FocusManager.push(Search)
+  | ActivityBar(ActivityBar.SearchClick) when isSearchOpen && !isSearchFocused =>
+    state |> FocusManager.push(Search)
+  | SearchHotkey when !isSearchOpen =>
+    {...state, pane: Pane.setOpen(Pane.Search)} |> FocusManager.push(Search)
+  | ActivityBar(ActivityBar.SearchClick) when !isSearchOpen =>
+    {...state, pane: Pane.setOpen(Pane.Search)} |> FocusManager.push(Search)
+  | StatusBar(StatusBarModel.DiagnosticsClicked) when isDiagnosticsOpen => {
+      ...state,
+      pane: Pane.setClosed(pane),
+    }
+  | DiagnosticsHotKey when isDiagnosticsOpen => {
+      ...state,
+      pane: Pane.setClosed(pane),
+    }
+  | StatusBar(StatusBarModel.DiagnosticsClicked) when !isDiagnosticsOpen => {
+      ...state,
+      pane: Pane.setOpen(Pane.Diagnostics),
+    }
+  | DiagnosticsHotKey when !isDiagnosticsOpen => {
+      ...state,
+      pane: Pane.setOpen(Pane.Diagnostics),
+    }
+  | _ => state
+  };
+};
```

Comment:
The control flow here is very confusing. Seems like it might be better to switch on `(action, pt)`, or use nested switches, instead of using guards, as that would give you exhaustiveness checking as well.

I'm also pretty sure that you're able to come up with a better name than `pt`...

Reply:
Made an attempt to clean it up! It's still a bit complex (it's essentially a state machine with 3 states - not focused/open, open but not focused, and then open but focused - and handling the transitions based on each action). Probably still a nicer way to express it! But I think it's better than it was.